File: lib/models/sos.dart
class TrustedContact {
  const TrustedContact({
    required this.id,
    required this.name,
    required this.phone,
    this.relationship,
  });

  final String id;
  final String name;
  final String phone;
  final String? relationship;

  String get maskedPhone {
    if (phone.length < 6) return phone;
    return '${phone.substring(0, 5)} ••• ${phone.substring(phone.length - 2)}';
  }

  factory TrustedContact.fromJson(Map<String, dynamic> json) => TrustedContact(
    id: json['id'] as String,
    name: json['name'] as String,
    phone: json['phone'] as String,
    relationship: json['relationship'] as String?,
  );
}

class SosReadiness {
  const SosReadiness({
    required this.ready,
    required this.contactCount,
    required this.smsProvider,
    required this.message,
  });

  final bool ready;
  final int contactCount;
  final String smsProvider;
  final String message;

  factory SosReadiness.fromJson(Map<String, dynamic> json) => SosReadiness(
    ready: json['ready'] as bool? ?? false,
    contactCount: json['contact_count'] as int? ?? 0,
    smsProvider: json['sms_provider'] as String? ?? 'disabled',
    message: json['message'] as String? ?? 'SOS unavailable',
  );
}

class SosResult {
  const SosResult({required this.deliveredTo, required this.contactCount});
  final int deliveredTo;
  final int contactCount;

  factory SosResult.fromJson(Map<String, dynamic> json) => SosResult(
    deliveredTo: json['delivered_to'] as int? ?? 0,
    contactCount: json['contact_count'] as int? ?? 0,
  );
}

File: backend/app/models/sos.py
from datetime import datetime, timezone
from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from app.models.incident import Location


class TrustedContactCreate(BaseModel):
    name: str = Field(min_length=2, max_length=100)
    phone: str = Field(pattern=r"^\+[1-9]\d{7,14}$", max_length=20)
    relationship: str | None = Field(default=None, max_length=50)


class TrustedContact(TrustedContactCreate):
    id: UUID = Field(default_factory=uuid4)
    user_id: UUID


class SosCreate(BaseModel):
    client_alert_id: UUID
    location: Location


class SosAlert(BaseModel):
    id: UUID = Field(default_factory=uuid4)
    user_id: UUID
    incident_id: UUID | None = None
    status: str = "active"
    delivered_to: int = 0
    contact_count: int = 0
    triggered_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class SosReadiness(BaseModel):
    ready: bool
    contact_count: int
    sms_provider: str
    message: str
